File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* ans;
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs(root, p, q);
        
        return ans;
    }
    
    int dfs(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        if(root == nullptr) return 0;
        
        int left = dfs(root->left, p, q);
        int right = dfs(root->right, p, q);
        
        if((left + right == 2 || (left + right == 1 && (root == p || root == q))) && ans == nullptr)
        {
            ans = root;
        }
        
        if(root == p || root == q)
        {
            return 1 + left + right;
        }
        else
        {
            return left + right;
        }
    }
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        ans = None
        
        
        def dfs(node):
            nonlocal ans
            
            if not node:
                return 0
            
            left = dfs(node.left)
            right = dfs(node.right)
            
            
            total = left + right
            if node == p or node == q:
                total += 1
            
            if total == 2 and ans == None:
                ans = node
            
            return total
        
        dfs(root)
        return ans